base/linalg/cholesky.jl

```diff
@@ -78,83 +104,149 @@ function chol!{T}(A::AbstractMatrix{T}, ::Type{LowerTriangular})
     return LowerTriangular(A)
 end
 
-"""
-    chol(A::AbstractMatrix) -> U
-
-Compute the Cholesky factorization of a positive definite matrix `A`
-and return the UpperTriangular matrix `U` such that `A = U'U`.
-"""
-function chol{T}(A::AbstractMatrix{T})
-    S = promote_type(typeof(chol(one(T))), Float32)
-    AA = similar(A, S, size(A))
-    copy!(AA, A)
-    chol!(AA)
-end
-function chol!(x::Number, uplo)
+## Numbers
+function _chol!(x::Number, uplo)
     rx = real(x)
     if rx != abs(x)
         throw(ArgumentError("x must be positive semidefinite"))
     end
     rxr = sqrt(rx)
     convert(promote_type(typeof(x), typeof(rxr)), rxr)
 end
+
+
+
+# chol!. Destructive methods for computing Cholesky factor of real symmetric of Hermitian
+# matrix
+chol!(A::Hermitian)                                = _chol!(A.data, UpperTriangular)
+chol!{T<:Real,S<:StridedMatrix}(A::Symmetric{T,S}) = _chol!(A.data, UpperTriangular)
+function chol!(A::StridedMatrix)
+    ishermitian(A) || throw(ArgumentError("matrix is not symmetric/Hermitian. Consider calling chol!(Hermitian(A)) instead."))
+    return _chol!(A, UpperTriangular)
+end
+
+
+
+# chol. Non-destructive methods for computing Cholesky factor of real symmetric of
+# Hermitian matrix. Promotes elements to type that is stable under sqrt roots.
+function chol(A::Hermitian)
+    T = promote_type(typeof(chol(one(eltype(A)))), Float32)
+    AA = similar(A, T, size(A))
+    copy!(AA, A.data)
+    chol!(Hermitian(AA))
+end
+function chol{T<:Real,S<:AbstractMatrix}(A::Symmetric{T,S})
+    TT = promote_type(typeof(chol(one(T))), Float32)
+    AA = similar(A, TT, size(A))
+    copy!(AA, A.data)
+    chol!(Hermitian(AA))
+end
+
+## for StridedMatrices, check that matrix is symmetric/Hermitian
+"""
+    chol(A) -> U
+
+Compute the Cholesky factorization of a positive definite matrix `A`
+and return the UpperTriangular matrix `U` such that `A = U'U`.
+"""
+function chol(A::StridedMatrix)
+    ishermitian(A) || throw(ArgumentError("matrix is not symmetric/Hermitian. Consider calling chol(Hermitian(A)) instead."))
+    return chol(Hermitian(A))
+end
+
+## Numbers
 """
```

Comment:
Often the documentation signatures cheat a bit for simplification. I think that having the `args...` part in the docstring would confuse more than help. Maybe you could write that "extra arguments are ignored". However, this is not part of the change in this PR so if we want that, it could be done in a separate PR.